coffea/nanoevents/schemas/physlite.py

```diff
@@ -17,31 +19,87 @@ class PHYSLITESchema(BaseSchema):
         super().__init__(base_form)
         self._form["contents"] = self._build_collections(self._form["contents"])
 
     def _build_collections(self, branch_forms):
         zip_groups = defaultdict(list)
         for key, ak_form in branch_forms.items():
+            # Normal fields
             key_fields = key.split("/")[-1].split(".")
             top_key = key_fields[0]
             sub_key = ".".join(key_fields[1:])
             objname = top_key.replace("Analysis", "").replace("AuxDyn", "")
-            zip_groups[objname].append((key, sub_key))
+            zip_groups[objname].append(((key, sub_key), ak_form))
+
+            # Global indices
+            for cross_references in [
+                self.cross_reference_indices,
+                self.cross_reference_elementlinks,
+            ]:
+                if (objname, sub_key) in cross_references:
+                    linkto_key = cross_references[(objname, sub_key)]
+                    if cross_references is self.cross_reference_indices:
+                        form = self._create_global_index_form(ak_form, key, linkto_key)
+                    elif cross_references is self.cross_reference_elementlinks:
+                        form = self._create_global_index_form_elementlink(
+                            ak_form, key, linkto_key
+                        )
+                    zip_groups[objname].append(((key, sub_key + "G"), form))
 
         contents = {}
-        for objname, keys in zip_groups.items():
+        for objname, keys_and_form in zip_groups.items():
             try:
                 contents[objname] = zip_forms(
-                    {sub_key: branch_forms[key] for key, sub_key in keys},
+                    {sub_key: form for (key, sub_key), form in keys_and_form},
                     objname,
                     self.mixins.get(objname, None),
                 )
             except NotImplementedError:
                 warnings.warn(f"Can't zip collection {objname}")
         return contents
 
+    @staticmethod
+    def _create_global_index_form(base_form, key, linkto_key):
+        form = copy.deepcopy(base_form)
+        form["content"]["form_key"] = quote(
+            f"{key},!load,{linkto_key},!load,!offsets,!to_numpy,!local2global"
+        )
+        form["content"]["itemsize"] = 8
+        form["content"]["primitive"] = "int64"
+        return form
+
+    @staticmethod
+    def _create_global_index_form_elementlink(base_form, key, linkto_key):
+        form = copy.deepcopy(base_form)
+        record = form["content"]["content"]["contents"]
+        form["content"]["content"] = record["m_persIndex"]
+        form["content"]["content"]["form_key"] = quote(
+            f"{key},!load,m_persIndex,!item,{linkto_key},!load,!offsets,!to_numpy,!local2global"
+        )
+        form["content"]["form_key"] = quote(
+            # the !data is a hack to avoid having the same key as the actual elementlink
+            f"{key},!load,!content,!data"
```

Comment:
i don't quite understand yet what's going on here - if i create a form where i e.g. have `events.Electrons.trackParticleLinks`  and `events.Electrons.trackParticleLinksG` which both have the same `form_key` on top level (which should be used only for the offsets), then the same array will come out, even if the `form_key` in the inner levels differs.

Reply:
Shouldn't the `trackParticleLinksG` form have a key that at least includes `!local2global` whereas the `trackParticleLinks` would not?

Reply:
yes, but since they both have one dimension more than the other fields of `Electrons` there is also a `form_key` on the level of the `ListOffsetArray`. So currently i have for `events.Electrons.layout.form`

```
{
    "class": "ListOffsetArray64",
    "offsets": "i64",
    "content": {
        "class": "RecordArray",
        "contents": {
            --------------------------------------
            <lots-of-single-jagged-fields>,
            --------------------------------------
            "trackParticleLinks": {
                "class": "VirtualArray",
                "form": {
                    "class": "ListOffsetArray64",
                    "offsets": "i64",
                    "content": {
                        "class": "RecordArray",
                        "contents": {
                            "m_persIndex": {
                                "class": "VirtualArray",
                                "form": {
                                    "class": "NumpyArray",
                                    "itemsize": 4,
                                    "format": "I",
                                    "primitive": "uint32",
                                    "form_key": "AnalysisElectronsAuxDyn.trackParticleLinks%2C%21load%2C%21content%2C%21content%2Cm_persIndex%2C%21item"
                                },
                                "has_length": true
                            },
                            "m_persKey" : <similar>
                       [...]
                    },
--->                "form_key": "AnalysisElectronsAuxDyn.trackParticleLinks%2C%21load%2C%21content"
                },
                "has_length": true
            },
            "trackParticleLinks__G__GSFTrackParticles": {
                "class": "VirtualArray",
                "form": {
                    "class": "ListOffsetArray64",
                    "offsets": "i64",
                    "content": {
                        "class": "NumpyArray",
                        "itemsize": 8,
                        "format": "l",
                        "primitive": "int64",
                        "form_key": "AnalysisElectronsAuxDyn.trackParticleLinks%2C%21load%2Cm_persIndex%2C%21item%2CGSFTrackParticlesAuxDyn.z0%2C%21load%2C%21offsets%2C%21local2global"
                    },
                    [...]
--->                "form_key": "AnalysisElectronsAuxDyn.trackParticleLinks%2C%21load%2C%21content%2C%21skip%2CGSFTrackParticlesAuxDyn.z0"
                },
                "has_length": true
            },
```
The form keys marked with `--->` are the ones i'm confused about. I would think they are only used for the offsets, but i'm not really sure how this works (there is no `!offset` after all? On the other hand there is this `"{attribute}"` thing in `ak.from_buffers` that sort of tells awkward this is supposed to be an offset?)

Probably i also should zip the global indices and the local indices together, then i wouldn't have that problem anymore ...

Reply:
> On the other hand there is this "{attribute}" thing in ak.from_buffers that sort of tells awkward this is supposed to be an offset?)

Right, for the higher level keys, ak.from_buffers is adding the `!offset` part automagically. So in principle, since both of these collections have the same offsets, it should be ok (desirable, even) that they have the same offsets key.

Reply:
maybe it's a bug - i opened scikit-hep/awkward-1.0#899